gobblin-config-management/gobblin-config-core/src/main/java/gobblin/config/common/impl/InMemoryValueInspector.java

```diff
@@ -0,0 +1,87 @@
+package gobblin.config.common.impl;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.WeakHashMap;
+
+import com.typesafe.config.Config;
+
+import gobblin.config.store.api.ConfigKeyPath;
+
+/**
+ * InMemoryValueInspector provide the caching layer for getting the {@com.typesafe.config.Config} from {@ConfigStore}
+ * 
+ * @author mitu
+ *
+ */
+public class InMemoryValueInspector implements ConfigStoreValueInspector{
+
+  private final ConfigStoreValueInspector valueFallback;
+
+  private final Map<ConfigKeyPath, Config> ownConfigMap;
+  private final Map<ConfigKeyPath, Config> recursiveConfigMap;
+
+  /**
+   * 
+   * @param valueFallback - the fall back {@ConfigStoreValueInspector} which used to get the raw {@com.typesafe.config.Config} 
+   * @param useStrongRef  - if true, use Strong reference in cache, else, use Weak reference in cache
+   */
+  public InMemoryValueInspector (ConfigStoreValueInspector valueFallback, boolean useStrongRef){
+    this.valueFallback = valueFallback;
+
+    if(useStrongRef){
+      this.ownConfigMap = new HashMap<>();
+      this.recursiveConfigMap = new HashMap<>();
+    }
+    else{
+      this.ownConfigMap = new WeakHashMap<>();
```

Comment:
You want to use soft values instead of weak values. Weak references get removed as soon as no other references exist in the JVM, soft references only get GCed if the memory is required.